### PR TITLE
Order of messages fix (#550)

### LINKED ISSUE #550: Order of messages
It is still possible to get different order of messages 

screen shot 2016-12-12 at 16 55 37

### PR DESCRIPTION
**Pull request by alwx**
Fixes #550.

:clock-value now works better and there should be no problems with message order. I've checked it — everything works fine.

### CODE REVIEW

**Comment by rasom** [comment]
1. send message 1 from A
2. send message 2 from B
3. turn on airplane mode on phone B
4. send message 3 and 4 from A
5. send message 5 from B
6. turn off airplane mode on B

A:
screen shot 2016-12-13 at 15 04 03

B:
img_0650

**Comment by alwx** [comment]
@rasom
This case is way more difficult.

Your internal clock-value is the same on both sides because one of the users is offline. As for me, this behaviour is more or less normal (the only problem here is that message appears in the middle of your messages). It happens because we cannot rely on timestamp and there is no middle man to control the message ordering and the only thing we can use is clock-value which, of course, can be out of sync in this case.

But there are two possible fixes here.
**First:** change the ordering to be 1 2 3 4 5 for user A and 1 2 5 3 4 for user B;

**Second:** wait for some time before sending a message and change the clock-value before sending. This will help to maintain the same ordering on both sides, but the UX may suffer. For example: 
1) you send a bunch of messages when you're offline and your buddy (he is online) sends something at the same time;
2) you become online;
3) we change clock-value of your messages and display your messages AFTER your buddy's messages.
I think in this case it may be difficult for you to find the buddy's messages because all these messages are above.

**Comment by alwx** [comment]
> But for now at least ensure that the order will be `1 2 3 4 5` for A and `1 2 5 3 4` for B.

Done.

**Comment by rasom** [comment]
Can reproduce `1 2 3 5 4`

screen shot 2016-12-17 at 12 20 06

img_0658

**Comment by rasom** [comment]
oh sorry @alwx, seems that branch wasn't updated locally, checking again

**Comment by rasom** [comment]
This time I'm completely sure that I was at `9baa098db168d60662b634ef7812e5e297d65beb` and can reproduce `1 2 3 5 4` :(

**Comment by alwx** [comment]
Done.
Now we maintain the same message order on both devices — so both users will see 1 2 3 4 5 in your example. It was hard, and I've tried several solutions before choosing this one.

There is one case when it is really difficult to keep the right ordering (and I don't try to) — when users send messages simultaneously. But I think it is okay — you will see your buddy's messages on the bottom in this case, and that's what most people expect.

File: src/status_im/chat/handlers/send_message.cljs [hunk 1]
```diff
@@ -179,22 +182,25 @@
...
-                         :content-type text-content-type
-                         :outgoing     true
-                         :timestamp    (time/now-ms)
-                         :clock-value  (inc clock-value)})
-            message'' (if group-chat
-                        (assoc message' :group-id chat-id :message-type :group-user-message)
-                        (assoc message' :to chat-id :message-type :user-message))
-            params'   (assoc params :message message'')]
+    (fn [{:keys [network-status]} [_ {:keys [chat-id identity message] :as params}]]
+      (let [{:keys [group-chat]} (get-in db [:chats chat-id])
```
Comment (rasom) [inline src/status_im/chat/handlers/send_message.cljs]: @alwx, I missed this when reviewed PR

> WARNING: Use of undeclared Var status-im.chat.handlers.send-message/db at line 186 /Users/rasom/clj/status-react/src/status_im/chat/handlers/send_message.cljs
WARNING: Use of undeclared Var status-im.chat.handlers.send-message/db at line 189 /Users/rasom/clj/status-react/src/status_im/chat/handlers/send_message.cljs

**Pull request merged**